providers/packet: respect bond interface requests

sdemos opened the pull request:
the metadata for network interfaces has a bond field. up until now, we ignored it and just always created a bond named bond0, attached all addresses to it, and bonded both physical interfaces to it. now, we create any bond interfaces mentioned in the network interfaces list, so there may be multiple bonds. it also doesn't configure bonds for interfaces that don't have a bond key listed in the metadata.

/cc @bgilbert

sdemos committed:
network: add Eq and PartialEq derives to structs

these structs are trivially comparable, so let's provide the
implementations.

sdemos committed:
providers/packet: respect bond interface requests

the metadata for network interfaces has a bond field. up until now, we
ignored it and just always created a bond named bond0, attached all
addresses to it, and bonded both physical interfaces to it. this worked
because that's how the network was setup on most packet instances, but
it breaks because on instance type 1E there is an interface that
shouldn't be a part of the bond.

now, we create any bond interfaces mentioned in the network interfaces
list, so there may be multiple bonds. it also doesn't configure bonds
for interfaces that don't have a bond key listed in the metadata, which
fixes the behavior on packet instance type 1E.

sdemos committed:
network: add unmanaged field to Interface

on packet, most of their machines have multiple nics. by default on most
instance types, these nics are bonded together. on some instance types,
particularly 1E, this is not the case, and instead, some interfaces are
not configured to use any bond devices. when this happens, the interface
should be configured to be Unmanaged. this commit adds the ability to
mark some interfaces as unmanaged. it also updates the packet provider
to configure an interface to be unmanaged if there are no bond devices
configured for it.

bgilbert commented on src/providers/packet/mod.rs:
```diff
@@ -117,83 +118,103 @@ fn get_dns_servers() -> Result<Vec<IpAddr>> {
+            let bond = Interface {
+                name: Some(bond_name.clone()),
+                priority: Some(5),
+                nameservers: get_dns_servers()?,
```
Strictly speaking, it's hypothetically possible for the return value to change between calls.

bgilbert commented on src/providers/packet/mod.rs:
```diff
@@ -117,83 +118,103 @@ fn get_dns_servers() -> Result<Vec<IpAddr>> {
+                ip_addresses: Vec::new(),
+                routes: Vec::new(),
+            };
+            if !bonds.contains(&bond) {
```
Won't this always return false because of `mac_address`?

sdemos replied:
it would do that. I didn't notice because it created another file of the same name with almost the same contents that overwrote the other one, so it still emitted the right number of bond devices

bgilbert commented on src/providers/packet/mod.rs:
```diff
@@ -117,83 +118,103 @@ fn get_dns_servers() -> Result<Vec<IpAddr>> {
+
+    // according to the folks from packet, all the addresses given to us in the
+    // network section should be attached to the first bond we find in the list
+    // of interfaces. there probably won't be no bonds in the list, but if we
```
clarify `won't be no bonds`

sdemos replied:
that sounded better in my head

bgilbert commented on src/providers/packet/mod.rs:
```diff
@@ -117,83 +118,103 @@ fn get_dns_servers() -> Result<Vec<IpAddr>> {
+
+    // remove panics if the index is out of bounds, but we know that there is at
+    // least one bond in the vector because we return if it's empty
+    let mut bond0 = bonds.remove(0);
```
There's nothing requiring it to actually be named `bond0`, so maybe e.g. `first_bond` would be more descriptive.

bgilbert commented on src/providers/packet/mod.rs:
```diff
@@ -117,83 +118,103 @@ fn get_dns_servers() -> Result<Vec<IpAddr>> {
-		("UpDelaySec".to_owned(), ".2".to_owned()),
-		("DownDelaySec".to_owned(), ".2".to_owned()),
-		("Mode".to_owned(), network::bonding_mode_to_string(&netinfo.bonding.mode)?),
+        ("TransmitHashPolicy".to_owned(), "layer3+4".to_owned()),
```
Inconsistent indentation?

sdemos replied:
fixed.

bgilbert reviewed (changes requested):
Please clean up the commit messages to provide as much detail as possible.  We should explicitly say that the no-bond case is for type 1E.

sdemos committed:
provider/packet: describe missing netif/state file

if coreos-metadata was run on a system where /run/systemd/netif/state
was missing, or if it failed to parse the file, it would return an error
without adding the context of what file it was trying to open or parse.
this chains that context onto any returned errors.
